pkg/server/auth/jwt.go

```diff
@@ -0,0 +1,123 @@
+	"net/http"
+	"strings"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"github.com/go-logr/logr"
+)
+
+// PrincipalGetter implementations are responsible for extracting a named
+// principal from an HTTP request.
+type PrincipalGetter interface {
+	// Principal extracts a principal from the http.Request.
+	// It's not an error for there to be no principal in the request.
+	Principal(r *http.Request) (*UserPrincipal, error)
+}
+
+// JWTCookiePrincipalGetter inspects a cookie for a JWT token
+// and returns a principal object.
+type JWTCookiePrincipalGetter struct {
+	log        logr.Logger
+	verifier   *oidc.IDTokenVerifier
+	cookieName string
+}
+
+func NewJWTCookiePrincipalGetter(log logr.Logger, verifier *oidc.IDTokenVerifier, cookieName string) PrincipalGetter {
+	return &JWTCookiePrincipalGetter{
+		log:        log,
+		verifier:   verifier,
+		cookieName: cookieName,
+	}
+}
+
+func (pg *JWTCookiePrincipalGetter) Principal(r *http.Request) (*UserPrincipal, error) {
+	pg.log.Info("attempt to read token from cookie")
+
+	cookie, err := r.Cookie(pg.cookieName)
+	if err == http.ErrNoCookie {
+		return nil, nil
+	}
+
+	return parseJWTToken(r.Context(), pg.verifier, cookie.Value)
+}
+
+// JWTAuthorizationHeaderPrincipalGetter inspects the Authorization
+// header (bearer token) for a JWT token and returns a principal
+// object.
+type JWTAuthorizationHeaderPrincipalGetter struct {
+	log      logr.Logger
+	verifier *oidc.IDTokenVerifier
+}
+
+func NewJWTAuthorizationHeaderPrincipalGetter(log logr.Logger, verifier *oidc.IDTokenVerifier) PrincipalGetter {
+	return &JWTAuthorizationHeaderPrincipalGetter{
+		log:      log,
+		verifier: verifier,
+	}
+}
+
+func (pg *JWTAuthorizationHeaderPrincipalGetter) Principal(r *http.Request) (*UserPrincipal, error) {
+	pg.log.Info("attempt to read token from auth header")
+
+	header := r.Header.Get("Authorization")
+	if header == "" {
+		return nil, nil
+	}
+
+	return parseJWTToken(r.Context(), pg.verifier, extractToken(header))
+}
+
+func extractToken(s string) string {
+	parts := strings.Split(s, " ")
+	if len(parts) != 2 {
+		return ""
+	}
+
+	if strings.TrimSpace(parts[0]) != "Bearer" {
+		return ""
+	}
+
+	return strings.TrimSpace(parts[1])
+}
+
+func parseJWTToken(ctx context.Context, verifier *oidc.IDTokenVerifier, rawIDToken string) (*UserPrincipal, error) {
+	token, err := verifier.Verify(ctx, rawIDToken)
+	if err != nil {
+		return nil, fmt.Errorf("failed to verify JWT token: %w", err)
+	}
+
+	var claims struct {
+		Email  string   `json:"email"`
+		Groups []string `json:"groups"`
+	}
+
+	if err := token.Claims(&claims); err != nil {
+		return nil, fmt.Errorf("failed to parse claims from the JWT token: %w", err)
+	}
+
+	return &UserPrincipal{ID: claims.Email, Groups: claims.Groups}, nil
+}
+
+// MultiAuth combines the JWTCookie and JWTAuthorizationHeader
+// principal getters together to look for a principal first in
+// a cookie and then in an Authorization header.
+type MultiAuthPrincipal []PrincipalGetter
```

Comment:
This comment doesn't seem quite right?

It doesn't seem to care about cookies or authorization ?

It loops over the provided authorizers and if it finds an error or a principal, it returns, otherwise it returns `nil, nil`.

Also, no tests for this?